File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第6章选数异或.cpp
//选数异或  2022年第十三届蓝桥杯省赛 C/C++大学A组、C组、研究生组，Java大学C组
//lanqiaoOJ 2081

#include<bits/stdc++.h>
using namespace std;
const int N = 100010 ;
int Left[N], pos[(1 << 20) + 10];
int a[N] ;
int tree[4*N];
int ls(int p){ return p<<1;  }     //左儿子
int rs(int p){ return p<<1|1;}     //右儿子
void push_up(int p){               //从下往上传递区间值
    tree[p] = max(tree[ls(p)], tree[rs(p)]);    //区间最大值
}
void build(int p, int pl, int pr) {
    if(pl == pr){
        tree[p] = Left[pl];
        return;
    }
    int mid = (pl + pr) >> 1;
    build(ls(p), pl, mid);
    build(rs(p), mid + 1, pr);
    push_up(p);
}
int query(int p, int pl, int pr, int L, int R){
    if(L <= pl && pr <= R)return tree[p];
    int mid = (pl + pr) >> 1;
    int res = 0;
    if(L <= mid) res = max(res, query(ls(p), pl, mid, L, R));
    if(R > mid)  res = max(res, query(rs(p), mid + 1, pr, L, R));
    return res;
}
int main(){
    int n, m, x;  cin >> n >> m >> x;
    for(int i = 1; i <= n; i++) {      //预处理Left数组
        scanf("%d",a+i);
        Left[i] = pos[a[i] ^ x];
        pos[a[i]] = i;
    }
    build(1, 1, n);
    while(m--){
        int L,R; scanf("%d%d",&L,&R);
        if(query(1, 1, n, L, R) >= L)  printf("yes\n");
        else                              printf("no\n");
    }
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第6章选数异或.py
#选数异或  2022年第十三届蓝桥杯省赛 C/C++大学A组、C组、研究生组，Java大学C组
#lanqiaoOJ 2081

N = 100010
Left=[0]*N
pos=[0]*((1 << 20) + 10)
tree = [0]*(N<<2)       #4倍
def build(p,pl,pr):
    if pl==pr:
        tree[p] = Left[pl]  
        return
    mid=(pl+pr)>>1
    build(p<<1,pl,mid)         #p<<1 是左儿子
    build(p<<1|1,mid+1,pr)     #p<<1|1是右儿子
    tree[p] = max(tree[p<<1],tree[p<<1|1])  #push_up
def query(p,pl,pr,L,R):
    if L<=pl and pr<=R:  return tree[p]
    mid=(pl+pr)>>1
    res = 0
    if L<=mid:  res=max(res,query(p<<1,pl,mid,L,R))
    if R>mid:   res=max(res,query(p<<1|1,mid+1,pr,L,R))
    return res
n, m, x = map(int,input().split())
a = list(input().split())
a.insert(0,0)              #加个a[0]。从a[1]开始
for i in range(1,n+1):
    a[i] = int(a[i])
    Left[i] = pos[a[i] ^ x]
    pos[a[i]] = i
build(1,1,n)  
for i in range(m):
    L, R = map(int,input().split())
    if query(1,1,n,L,R)>=L: print('yes')
    else:                   print('no')
